File: finance-app/src/app/services/portfolio.service.ts
import { Injectable } from '@angular/core';
import { Observable, combineLatest, map, catchError, of } from 'rxjs';
import { Portfolio, DashboardSummary, AssetType, TransactionType } from '../models/asset.model';
import { AssetService } from './asset.service';
import { TransactionService } from './transaction.service';
import { ApiService } from './api.service';
import { ExchangeRateService } from './exchange-rate.service';
import { FundTaxRateService } from './fund-tax-rate.service';

export interface HistoricalInvestment {
  assetId: string;
  assetSymbol: string;
  assetName: string;
  asset?: any;
  totalBought: number;
  totalSold: number;
  averageBuyPrice: number;
  averageSellPrice: number;
  realizedProfitLoss: number;
  realizedProfitLossPercentage: number;
  transactions: any[];
}

export interface EmailSummaryRequest {
  recipientEmail: string;
  subject?: string;
  message?: string;
  userId?: string;
}

@Injectable({
  providedIn: 'root',
})
export class PortfolioService {
  constructor(
    private assetService: AssetService,
    private transactionService: TransactionService,
    private apiService: ApiService,
    private exchangeRateService: ExchangeRateService,
    private fundTaxRateService: FundTaxRateService
  ) {}

  // Backend'de portfolio endpoint'i hazır olduğunda bu metodu kullanacağız
  private getPortfolioFromApi(): Observable<Portfolio[]> {
    return this.apiService.get<Portfolio[]>('/portfolio').pipe(
      catchError((error) => {
        console.error('Portfolio API call failed:', error);
        return of([]);
      })
    );
  }

  // Backend'de dashboard endpoint'i hazır olduğunda bu metodu kullanacağız
  private getDashboardFromApi(): Observable<DashboardSummary> {
    return this.apiService.get<DashboardSummary>('/dashboard').pipe(
      catchError((error) => {
        console.error('Dashboard API call failed:', error);
        return this.getDashboardSummary();
      })
    );
  }

  emailDashboardSummary(request: EmailSummaryRequest): Observable<void> {
    const body = {
      recipientEmail: request.recipientEmail,
      subject: request.subject,
      message: request.message,
      userId: request.userId ?? 'default-user',
    };

    return this.apiService.post<void>('/portfolio/email-summary', body);
  }

  getPortfolio(): Observable<Portfolio[]> {
    return combineLatest([
      this.assetService.getAssets(),
      this.transactionService.getTransactions(),
      this.fundTaxRateService.getRateMap(),
    ]).pipe(
      map(([assets, transactions, fundTaxRates]) => {
        const portfolioMap = new Map<string, Portfolio>();

        // Group transactions by asset
        transactions.forEach((transaction) => {
          let asset = null;
          let portfolioKey = '';

          if (transaction.assetId) {
            // Normal asset with ID
            asset = assets.find((a) => a.id === transaction.assetId);
            if (!asset) return;
            portfolioKey = transaction.assetId;
          } else {
            // Generic asset (Gold, FixedDeposit) without specific asset ID
            // Create a virtual asset based on transaction type
            const assetType = this.getAssetTypeFromTransaction(transaction);
            if (!assetType) return;

            portfolioKey = `generic_${assetType}`;
            asset = this.createVirtualAsset(assetType);
          }

          if (!portfolioMap.has(portfolioKey)) {
            portfolioMap.set(portfolioKey, {
              assetId: portfolioKey,
              assetSymbol: asset.symbol,
              assetName: asset.name,
              asset: asset,
              totalQuantity: 0,
              quantity: 0,
              averagePrice: 0,
              currentValue: 0,
              totalCost: 0,
              grossProfitLoss: 0,
              grossProfitLossPercentage: 0,
              withholdingTaxRatePercent: 0,
              netProfitLoss: 0,
              netProfitLossPercentage: 0,
              profitLoss: 0,
              profitLossPercentage: 0,
              currency: asset.currency,
              lastUpdated: new Date(),
              transactions: [],
            });
          }

          const portfolio = portfolioMap.get(portfolioKey)!;
          portfolio.transactions.push(transaction);

          // Calculate portfolio metrics
          if (
            transaction.type === TransactionType.BUY ||
            transaction.type === TransactionType.DEPOSIT_ADD ||
            transaction.type === TransactionType.DEPOSIT_INCOME
          ) {
            const totalCostBefore = portfolio.totalCost;
            const totalQuantityBefore = portfolio.totalQuantity;

            portfolio.totalQuantity += transaction.quantity;

            // Faiz geliri için cost'a eklemeyelim
            if (transaction.type !== TransactionType.DEPOSIT_INCOME) {
              portfolio.totalCost += transaction.price * transaction.quantity + (transaction.fees || 0);
            }

            // Recalculate average price
            if (portfolio.totalCost > 0) {
              portfolio.averagePrice = portfolio.totalCost / portfolio.totalQuantity;
            }
          } else if (
            transaction.type === TransactionType.SELL ||
            transaction.type === TransactionType.DEPOSIT_WITHDRAW
          ) {
            // SELL veya DEPOSIT_WITHDRAW
            portfolio.totalQuantity -= transaction.quantity;
            // For sells, we don't adjust total cost as it represents what we paid
          }
        });

        // Calculate current values, profit/loss
        const portfolios = Array.from(portfolioMap.values()).filter((p) => p.totalQuantity > 0);

        portfolios.forEach((portfolio) => {
          portfolio.currentValue = portfolio.totalQuantity * portfolio.asset!.currentPrice;

          const grossProfitLoss = portfolio.currentValue - portfolio.totalCost;
          const grossProfitLossPercentage = portfolio.totalCost > 0 ? (grossProfitLoss / portfolio.totalCost) * 100 : 0;

          const assetType = portfolio.asset?.type;
          const withholdingRatePercent =
            assetType === AssetType.Fund ? Number(fundTaxRates.get(portfolio.asset!.id) ?? 0) : 0;

          const netProfitLoss = this.applyWithholdingTax(grossProfitLoss, assetType, withholdingRatePercent);
          const netProfitLossPercentage = portfolio.totalCost > 0 ? (netProfitLoss / portfolio.totalCost) * 100 : 0;

          portfolio.grossProfitLoss = grossProfitLoss;
          portfolio.grossProfitLossPercentage = grossProfitLossPercentage;
          portfolio.withholdingTaxRatePercent = withholdingRatePercent;
          portfolio.netProfitLoss = netProfitLoss;
          portfolio.netProfitLossPercentage = netProfitLossPercentage;

          // UI: profitLoss alanı fonlar için net, diğerleri için brüt
          portfolio.profitLoss = assetType === AssetType.Fund ? netProfitLoss : grossProfitLoss;
          portfolio.profitLossPercentage =
            assetType === AssetType.Fund ? netProfitLossPercentage : grossProfitLossPercentage;
          portfolio.quantity = portfolio.totalQuantity; // quantity ile totalQuantity aynı olsun
        });

        return portfolios;
      })
    );
  }

  private applyWithholdingTax(grossProfitLoss: number, assetType: AssetType | undefined, ratePercent: number): number {
    if (assetType !== AssetType.Fund) {
      return grossProfitLoss;
    }

    // zarar varsa stopaj uygulanmaz
    if (grossProfitLoss <= 0) {
      return grossProfitLoss;
    }

    const rate = Math.min(100, Math.max(0, Number(ratePercent ?? 0))) / 100;
    return grossProfitLoss * (1 - rate);
  }

  getHistoricalInvestments(): Observable<HistoricalInvestment[]> {
    return combineLatest([this.assetService.getAssets(), this.transactionService.getTransactions()]).pipe(
      map(([assets, transactions]) => {
        const historicalMap = new Map<string, HistoricalInvestment>();

        // Group transactions by asset
        transactions.forEach((transaction) => {
          const asset = assets.find((a) => a.id === transaction.assetId);
          if (!asset) return;

          if (!historicalMap.has(transaction.assetId)) {
            historicalMap.set(transaction.assetId, {
              assetId: transaction.assetId,
              assetSymbol: asset.symbol,
              assetName: asset.name,
              asset: asset,
              totalBought: 0,
              totalSold: 0,
              averageBuyPrice: 0,
              averageSellPrice: 0,
              realizedProfitLoss: 0,
              realizedProfitLossPercentage: 0,
              transactions: [],
            });
          }

          const historical = historicalMap.get(transaction.assetId)!;
          historical.transactions.push(transaction);
        });

        // Calculate historical metrics
        const historicalInvestments = Array.from(historicalMap.values());

        historicalInvestments.forEach((historical) => {
          const buyTransactions = historical.transactions.filter((t) => t.type === TransactionType.BUY);
          const sellTransactions = historical.transactions.filter((t) => t.type === TransactionType.SELL);

          // Calculate totals
          historical.totalBought = buyTransactions.reduce((sum, t) => sum + t.quantity, 0);
          historical.totalSold = sellTransactions.reduce((sum, t) => sum + t.quantity, 0);

          // Calculate average prices
          if (buyTransactions.length > 0) {
            const totalBuyValue = buyTransactions.reduce((sum, t) => sum + t.quantity * t.price, 0);
            historical.averageBuyPrice = totalBuyValue / historical.totalBought;
          }

          if (sellTransactions.length > 0) {
            const totalSellValue = sellTransactions.reduce((sum, t) => sum + t.quantity * t.price, 0);
            historical.averageSellPrice = totalSellValue / historical.totalSold;
          }

          // Calculate realized profit/loss
          const soldQuantity = Math.min(historical.totalBought, historical.totalSold);
          if (soldQuantity > 0 && historical.averageBuyPrice > 0 && historical.averageSellPrice > 0) {
            const costBasis = soldQuantity * historical.averageBuyPrice;
            const saleValue = soldQuantity * historical.averageSellPrice;
            historical.realizedProfitLoss = saleValue - costBasis;
            historical.realizedProfitLossPercentage = (historical.realizedProfitLoss / costBasis) * 100;
          }
        });

        // Only return assets that have been completely sold (totalSold >= totalBought)
        return historicalInvestments.filter((h) => h.totalSold >= h.totalBought && h.totalSold > 0);
      })
    );
  }

  getTotalProfitLoss(): Observable<{ totalPL: number; realizedPL: number; unrealizedPL: number }> {
    return combineLatest([this.getPortfolio(), this.getHistoricalInvestments()]).pipe(
      map(([currentPortfolio, historicalInvestments]) => {
        const unrealizedPL = currentPortfolio.reduce((sum, p) => sum + p.profitLoss, 0);
        const realizedPL = historicalInvestments.reduce((sum, h) => sum + h.realizedProfitLoss, 0);
        const totalPL = unrealizedPL + realizedPL;

        return {
          totalPL,
          realizedPL,
          unrealizedPL,
        };
      })
    );
  }

  getPortfolioByType(type: AssetType): Observable<Portfolio[]> {
    return this.getPortfolio().pipe(
      map((portfolios) => {
        return portfolios.filter((p) => {
          // Normal asset'ler için type kontrolü
          if (p.asset?.type === type) {
            return true;
          }

          // Virtual asset'ler için özel kontrol
          if (type === AssetType.FixedDeposit && p.assetId === 'generic_5') {
            return true;
          }

          if (type === AssetType.Gold && p.assetId === 'generic_2') {
            return true;
          }

          return false;
        });
      })
    );
  }

  getHistoricalInvestmentsByType(type: AssetType): Observable<HistoricalInvestment[]> {
    return this.getHistoricalInvestments().pipe(
      map((historicals) => historicals.filter((h) => h.asset?.type === type))
    );
  }

  getDashboardSummary(): Observable<DashboardSummary> {
    const selectedCurrency = this.exchangeRateService.getSelectedCurrency();

    return this.getPortfolio().pipe(
      map((portfolios) => {
        // Para birimi dönüşümü UI'da cache'deki exchange rate'leri kullanarak yapılır
        const convertedPortfolios = portfolios.map((portfolio) => {
          const convertedCurrentValue = this.exchangeRateService.convertToSelectedCurrencyWithCache(
            portfolio.currentValue,
            portfolio.currency
          );
          const convertedTotalCost = this.exchangeRateService.convertToSelectedCurrencyWithCache(
            portfolio.totalCost,
            portfolio.currency
          );

          return {
            ...portfolio,
            convertedCurrentValue,
            convertedTotalCost,
          };
        });

        // Toplam hesaplamalar
        const totalValue = convertedPortfolios.reduce((sum, p) => sum + p.convertedCurrentValue, 0);
        const totalCost = convertedPortfolios.reduce((sum, p) => sum + p.convertedTotalCost, 0);
        const totalProfitLoss = totalValue - totalCost;
        const totalProfitLossPercentage = totalCost > 0 ? (totalProfitLoss / totalCost) * 100 : 0;

        // Asset type'a göre gruplama (dönüştürülmüş değerlerle)
        const portfoliosByType = new Map<AssetType, Portfolio[]>();
        convertedPortfolios.forEach((portfolio) => {
          if (portfolio.asset) {
            const type = portfolio.asset.type;
            if (!portfoliosByType.has(type)) {
              portfoliosByType.set(type, []);
            }

            // Portfolio'ya dönüştürülmüş değerleri ekle (fonlarda net kâr/zarar korunur)
            const convertedGrossProfitLoss = portfolio.convertedCurrentValue - portfolio.convertedTotalCost;
            const convertedGrossProfitLossPercentage =
              portfolio.convertedTotalCost > 0 ? (convertedGrossProfitLoss / portfolio.convertedTotalCost) * 100 : 0;

            const withholdingRatePercent = Number((portfolio as any).withholdingTaxRatePercent ?? 0);
            const convertedNetProfitLoss = this.applyWithholdingTax(
              convertedGrossProfitLoss,
              type,
              withholdingRatePercent
            );
            const convertedNetProfitLossPercentage =
              portfolio.convertedTotalCost > 0 ? (convertedNetProfitLoss / portfolio.convertedTotalCost) * 100 : 0;

            const convertedProfitLoss = type === AssetType.Fund ? convertedNetProfitLoss : convertedGrossProfitLoss;
            const convertedProfitLossPercentage =
              type === AssetType.Fund ? convertedNetProfitLossPercentage : convertedGrossProfitLossPercentage;

            const portfolioWithConversion = {
              ...portfolio,
              currentValue: portfolio.convertedCurrentValue,
              totalCost: portfolio.convertedTotalCost,
              grossProfitLoss: convertedGrossProfitLoss,
              grossProfitLossPercentage: convertedGrossProfitLossPercentage,
              netProfitLoss: convertedNetProfitLoss,
              netProfitLossPercentage: convertedNetProfitLossPercentage,
              profitLoss: convertedProfitLoss,
              profitLossPercentage: convertedProfitLossPercentage,
              currency: selectedCurrency,
            };
            portfoliosByType.get(type)!.push(portfolioWithConversion);
          }
        });

        return {
          totalValue,
          totalCost,
          totalProfitLoss,
          totalProfitLossPercentage,
          portfoliosByType,
          displayCurrency: selectedCurrency,
        };
      })
    );
  }

  // Helper fonksiyonlar
  private getAssetTypeFromTransaction(transaction: any): AssetType | null {
    // Transaction'ın type'ına göre asset type'ını belirle
    if (
      transaction.type === TransactionType.DEPOSIT_ADD ||
      transaction.type === TransactionType.DEPOSIT_WITHDRAW ||
      transaction.type === TransactionType.DEPOSIT_INCOME
    ) {
      return AssetType.FixedDeposit;
    }

    // Diğer generic asset type'lar için gelecekte eklenebilir
    // Şimdilik Gold için özel bir logic yok, çünkü Gold da normal asset olarak ekleniyor

    return null;
  }

  private createVirtualAsset(assetType: AssetType): any {
    switch (assetType) {
      case AssetType.FixedDeposit:
        return {
          id: 'virtual_fixed_deposit',
          symbol: 'DEPOSIT',
          name: 'Vadeli Mevduat',
          type: AssetType.FixedDeposit,
          currentPrice: 1, // Vadeli mevduat için fiyat 1 (TRY bazında)
          currency: 'TRY',
          lastUpdated: new Date(),
          changePercentage: 0,
          changeValue: 0,
        };
      case AssetType.Gold:
        return {
          id: 'virtual_gold',
          symbol: 'GOLD',
          name: 'Altın',
          type: AssetType.Gold,
          currentPrice: 2500, // Default gold price - gerçek fiyat API'den alınmalı
          currency: 'TRY',
          lastUpdated: new Date(),
          changePercentage: 0,
          changeValue: 0,
        };
      default:
        return null;
    }
  }
}

File: finance-api/finance-api/Services/PortfolioService.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using FinanceApi.Data;
using FinanceApi.Models;
using FinanceApi.Models.Dtos;
using Microsoft.EntityFrameworkCore;

namespace FinanceApi.Services
{
    public class PortfolioService : IPortfolioService
    {
        private readonly FinanceDbContext _context;
        private readonly IExchangeRateService _exchangeRateService;

        public PortfolioService(FinanceDbContext context, IExchangeRateService exchangeRateService)
        {
            _context = context;
            _exchangeRateService = exchangeRateService;
        }

        public async Task<IEnumerable<PortfolioDto>> GetUserPortfolioAsync(string userId)
        {
            var fundTaxRates = await GetFundTaxRateMapAsync(userId);

            var portfolios = await _context.Portfolios
                .AsNoTracking()
                .Include(p => p.Asset)
                .Where(p => p.UserId == userId && p.TotalQuantity > 0)
                .OrderBy(p => p.Asset.Symbol)
                .ToListAsync();

            return portfolios.Select(p => MapPortfolio(p, fundTaxRates)).ToList();
        }

        public async Task<IEnumerable<PortfolioDto>> GetUserPortfolioByTypeAsync(string userId, AssetType type)
        {
            var fundTaxRates = await GetFundTaxRateMapAsync(userId);

            var portfolios = await _context.Portfolios
                .AsNoTracking()
                .Include(p => p.Asset)
                .Where(p => p.UserId == userId && p.TotalQuantity > 0 && p.Asset.Type == type)
                .OrderBy(p => p.Asset.Symbol)
                .ToListAsync();

            return portfolios.Select(p => MapPortfolio(p, fundTaxRates)).ToList();
        }

        public async Task<PortfolioDto?> GetUserPortfolioByAssetAsync(string userId, string assetId)
        {
            var portfolio = await _context.Portfolios
                .AsNoTracking()
                .Include(p => p.Asset)
                .FirstOrDefaultAsync(p => p.UserId == userId && p.AssetId == assetId);

            if (portfolio == null)
            {
                return null;
            }

            var fundTaxRates = await GetFundTaxRateMapAsync(userId);
            return MapPortfolio(portfolio, fundTaxRates);
        }

        public async Task<DashboardSummaryDto> GetDashboardSummaryAsync(string userId, string targetCurrency = "TRY")
        {
            var normalizedCurrency = string.IsNullOrWhiteSpace(targetCurrency)
                ? "TRY"
                : targetCurrency.ToUpperInvariant();

            var fundTaxRates = await GetFundTaxRateMapAsync(userId);

            var portfolioEntities = await _context.Portfolios
                .AsNoTracking()
                .Include(p => p.Asset)
                .Where(p => p.UserId == userId && p.TotalQuantity > 0)
                .ToListAsync();

            var portfolioDtos = portfolioEntities.Select(p => MapPortfolio(p, fundTaxRates)).ToList();

            if (portfolioDtos.Any())
            {
                await ConvertPortfoliosToCurrencyAsync(portfolioDtos, normalizedCurrency);
            }

            // ProfitLoss: net P&L (fonlarda stopaj sonrası)
            var totalValue = portfolioDtos.Sum(p => p.CurrentValue);
            var totalCost = portfolioDtos.Sum(p => p.TotalCost);
            var totalProfitLoss = portfolioDtos.Sum(p => p.ProfitLoss);
            var totalProfitLossPercentage = totalCost > 0 ? (totalProfitLoss / totalCost) * 100 : 0;

            var portfoliosByType = portfolioDtos
                .GroupBy(p => p.AssetType)
                .ToDictionary(g => g.Key, g => g.ToList());

            var lastUpdated = portfolioDtos.Any()
                ? portfolioDtos.Max(p => p.LastUpdated)
                : DateTime.UtcNow;

            return new DashboardSummaryDto
            {
                TotalValue = totalValue,
                TotalCost = totalCost,
                TotalProfitLoss = totalProfitLoss,
                TotalProfitLossPercentage = totalProfitLossPercentage,
                AssetCount = portfolioDtos.Count,
                PortfoliosByType = portfoliosByType,
                LastUpdated = lastUpdated,
                Currency = normalizedCurrency
            };
        }

        public async Task<IEnumerable<AssetTypePerformanceDto>> GetAssetTypePerformanceAsync(string userId)
        {
            var fundTaxRates = await GetFundTaxRateMapAsync(userId);

            var portfolios = await _context.Portfolios
                .AsNoTracking()
                .Include(p => p.Asset)
                .Where(p => p.UserId == userId && p.TotalQuantity > 0)
                .ToListAsync();

            var portfolioDtos = portfolios.Select(p => MapPortfolio(p, fundTaxRates)).ToList();

            return portfolioDtos
                .GroupBy(p => p.AssetType)
                .Select(g => new AssetTypePerformanceDto
                {
                    AssetType = g.Key,
                    TotalValue = g.Sum(p => p.CurrentValue),
                    TotalCost = g.Sum(p => p.TotalCost),
                    TotalProfitLoss = g.Sum(p => p.ProfitLoss),
                    TotalProfitLossPercentage = g.Sum(p => p.TotalCost) > 0
                        ? (g.Sum(p => p.ProfitLoss) / g.Sum(p => p.TotalCost)) * 100
                        : 0,
                    AssetCount = g.Count()
                })
                .ToList();
        }

        public async Task RecalculatePortfolioAsync(string userId, string assetId)
        {
            var portfolio = await _context.Portfolios
                .FirstOrDefaultAsync(p => p.UserId == userId && p.AssetId == assetId);

            if (portfolio == null)
            {
                return;
            }

            // Basit ortalama maliyet: tüm BUY işlemlerinin ortalaması.
            // SELL işlemleri sadece quantity azaltır.
            var buyTransactions = await _context.Transactions
                .Where(t => t.AssetId == assetId && t.UserId == userId && t.Type == TransactionType.BUY)
                .ToListAsync();

            var sellTransactions = await _context.Transactions
                .Where(t => t.AssetId == assetId && t.UserId == userId && t.Type == TransactionType.SELL)
                .ToListAsync();

            if (buyTransactions.Any())
            {
                var totalBuyQuantity = buyTransactions.Sum(t => t.Quantity);
                var totalSellQuantity = sellTransactions.Sum(t => t.Quantity);
                var totalCost = buyTransactions.Sum(t => t.Quantity * t.Price);

                portfolio.TotalQuantity = totalBuyQuantity - totalSellQuantity;
                portfolio.AveragePrice = totalBuyQuantity > 0 ? totalCost / totalBuyQuantity : 0;
            }
            else
            {
                portfolio.TotalQuantity = 0;
                portfolio.AveragePrice = 0;
            }

            if (portfolio.TotalQuantity <= 0)
            {
                _context.Portfolios.Remove(portfolio);
            }

            portfolio.LastUpdated = DateTime.UtcNow;
            await _context.SaveChangesAsync();
        }

        public async Task RecalculateAllPortfoliosAsync(string userId)
        {
            var portfolios = await _context.Portfolios
                .Where(p => p.UserId == userId)
                .ToListAsync();

            foreach (var portfolio in portfolios)
            {
                await RecalculatePortfolioAsync(userId, portfolio.AssetId);
            }
        }

        private async Task ConvertPortfoliosToCurrencyAsync(List<PortfolioDto> portfolios, string targetCurrency)
        {
            var rateCache = new Dictionary<string, decimal>(StringComparer.OrdinalIgnoreCase);

            foreach (var portfolio in portfolios)
            {
                var sourceCurrency = string.IsNullOrWhiteSpace(portfolio.Currency)
                    ? targetCurrency
                    : portfolio.Currency.ToUpperInvariant();

                if (sourceCurrency == targetCurrency)
                {
                    portfolio.Currency = targetCurrency;
                    continue;
                }

                var currentPrice = await ConvertWithCacheAsync(portfolio.CurrentPrice, sourceCurrency, targetCurrency, rateCache);
                var averagePrice = await ConvertWithCacheAsync(portfolio.AveragePrice, sourceCurrency, targetCurrency, rateCache);

                portfolio.CurrentPrice = Math.Round(currentPrice, 4);
                portfolio.AveragePrice = Math.Round(averagePrice, 4);

                portfolio.TotalCost = Math.Round(portfolio.TotalQuantity * portfolio.AveragePrice, 2);
                portfolio.CurrentValue = Math.Round(portfolio.TotalQuantity * portfolio.CurrentPrice, 2);

                var grossProfitLoss = Math.Round(portfolio.CurrentValue - portfolio.TotalCost, 2);
                var grossProfitLossPercentage = portfolio.TotalCost > 0
                    ? Math.Round((grossProfitLoss / portfolio.TotalCost) * 100, 2)
                    : 0;

                var netProfitLoss = ApplyWithholdingTax(grossProfitLoss, portfolio.AssetType, portfolio.WithholdingTaxRatePercent);
                var netProfitLossPercentage = portfolio.TotalCost > 0
                    ? Math.Round((netProfitLoss / portfolio.TotalCost) * 100, 2)
                    : 0;

                portfolio.GrossProfitLoss = grossProfitLoss;
                portfolio.GrossProfitLossPercentage = grossProfitLossPercentage;
                portfolio.NetProfitLoss = netProfitLoss;
                portfolio.NetProfitLossPercentage = netProfitLossPercentage;

                // Back-compat: UI halen ProfitLoss alanını kullanıyor (net)
                portfolio.ProfitLoss = netProfitLoss;
                portfolio.ProfitLossPercentage = netProfitLossPercentage;

                portfolio.Currency = targetCurrency;
            }
        }

        private async Task<decimal> ConvertWithCacheAsync(
            decimal amount,
            string fromCurrency,
            string toCurrency,
            IDictionary<string, decimal> cache)
        {
            if (amount == 0m || fromCurrency == toCurrency)
            {
                return amount;
            }

            var cacheKey = $"{fromCurrency}->{toCurrency}";

            if (!cache.TryGetValue(cacheKey, out var rate))
            {
                var convertedUnit = await _exchangeRateService.ConvertCurrencyAsync(1m, fromCurrency, toCurrency);
                if (convertedUnit == 0m)
                {
                    return amount;
                }

                rate = convertedUnit;
                cache[cacheKey] = rate;
            }

            return amount * rate;
        }

        private static decimal ApplyWithholdingTax(decimal grossProfitLoss, AssetType assetType, decimal ratePercent)
        {
            if (assetType != AssetType.Fund)
            {
                return grossProfitLoss;
            }

            // Zarar varsa stopaj yok.
            if (grossProfitLoss <= 0m)
            {
                return grossProfitLoss;
            }

            var rate = Math.Clamp(ratePercent, 0m, 100m) / 100m;
            return grossProfitLoss * (1m - rate);
        }

        private async Task<Dictionary<string, decimal>> GetFundTaxRateMapAsync(string userId)
        {
            return await _context.FundTaxRates
                .AsNoTracking()
                .Where(x => x.UserId == userId)
                .ToDictionaryAsync(x => x.AssetId, x => x.RatePercent);
        }

        private PortfolioDto MapPortfolio(Portfolio portfolio, IReadOnlyDictionary<string, decimal> fundTaxRates)
        {
            var totalCost = portfolio.TotalCost;
            var currentValue = portfolio.CurrentValue;

            var grossProfitLoss = currentValue - totalCost;
            var grossProfitLossPercentage = totalCost > 0 ? (grossProfitLoss / totalCost) * 100 : 0;

            var rate = 0m;
            if (portfolio.Asset.Type == AssetType.Fund && fundTaxRates.TryGetValue(portfolio.AssetId, out var configuredRate))
            {
                rate = configuredRate;
            }

            var netProfitLoss = ApplyWithholdingTax(grossProfitLoss, portfolio.Asset.Type, rate);
            var netProfitLossPercentage = totalCost > 0 ? (netProfitLoss / totalCost) * 100 : 0;

            return new PortfolioDto
            {
                Id = portfolio.Id,
                AssetId = portfolio.AssetId,
                AssetSymbol = portfolio.Asset.Symbol,
                AssetName = portfolio.Asset.Name,
                AssetType = portfolio.Asset.Type,
                TotalQuantity = portfolio.TotalQuantity,
                AveragePrice = portfolio.AveragePrice,
                CurrentPrice = portfolio.Asset.CurrentPrice,
                CurrentValue = currentValue,
                TotalCost = totalCost,

                GrossProfitLoss = grossProfitLoss,
                GrossProfitLossPercentage = grossProfitLossPercentage,
                WithholdingTaxRatePercent = rate,
                NetProfitLoss = netProfitLoss,
                NetProfitLossPercentage = netProfitLossPercentage,

                ProfitLoss = netProfitLoss,
                ProfitLossPercentage = netProfitLossPercentage,

                Currency = portfolio.Asset.Currency,
                LastUpdated = portfolio.LastUpdated
            };
        }
    }
}
